Layer documentation enhancements (#125)
napari/docs: docs/howtos/layers/shapes.md

lines. You will also understand how to add a `shapes` layer and edit it from the
GUI and from the console.

For more information about layers, refer to [Layers at a glance](../../guides/layers).

## When to use the `shapes` layer

The `shapes` layer allows you to display a list of an NxD arrays, where each
array corresponds to one shape, specified by N points in D coordinates. You can
adjust the position, size, face color, edge color, and opacity of all the shapes
independently, both programmatically and from the GUI.

## Controlling the `shapes` layer using the GUI

Before we start talking about all the controls on the GUI, here is an overview
of how to handle the `shapes` layer.

### Selecting, resizing, moving, editing, and deleting shapes

All shapes are edited in the same way:
1. Click the select shapes tool.
2. Draw a box around the shape you want to edit.
3. Adjust the size or contour of the shape using the square handles that appear
   on the bounding box of the shape. When resizing a layer, hold down the
   `shift` key  to lock the aspect ratio of the shape. (You do not have to hold
   the shift key down.) Then you can continue to resize the shape with a fixed

>    aspect ratio. Here is a shape being resized:

Reviewer: ```suggestion
   `shift` key  to lock the aspect ratio of the shape. Then you can continue
    to resize the shape with a fixed aspect ratio. Note: you have to hold
   the shift key down and when finished release the mouse button first!
    Here is a shape being resized:
```

  the rest. Once a vertex is selected, it can be moved by dragging it to a new
  location.

* **Select shapes**
  ![image: Select shapes](../../images/shape-select-shape.png)

  Use this tool to select the shape or shapes you want to delete. To select more
  than one shape, select one, hold down the shift key and select the rest.

* **Pan/zoom**
  ![image: Select shapes](../../images/pan-zoom-tool.png)

  Use this tool to pan around the image or zoom in. Pan and zoom functionality
  is disabled when using the adding and editing tools. Temporarily re-enable pan
  and zoom by pressing and holding the spacebar. This feature can be useful if
  you want to move around the shapes layer as you edit it.

* **Move to back**
  ![image: Move to back](../../images/shape-move-to-back.png)

  Using the `select shapes` tool, select the shape to move behind other shapes
  and then click this button.

* **Move to front**
  ![image: Move to front](../../images/shape-move-to-front.png)

  Using the `select shapes` tool, select the shape to move in front of other
  shapes and then click this button.

* **Add ellipses**
  ![image: Add ellipses](../../images/shape-add-ellipses.png)

>   Use this tool to draw ovals or circles. Click the tool, click the point where

Reviewer: ```suggestion
  Use this tool to draw ovals or circles. Click the tool (or press `e`, 
  the default keybinding), click the point where
```

  you want the ellipse to begin. Hold down the left button on the mouse and go
  to the point where you want the ellipse to end. Release the mouse button. Add
  other ellipses as needed. A single click creates an ellipse of default size
  centered on that click.

* **Add rectangles**
  ![image: Add rectangles](../../images/shape-add-rectangles.png)

  Select `Add rectangles` from the layer controls panel or by pressing the `R`
  key when the `shapes` layer is selected. Click and drag the rectangle to the
  desired size. Releasing the mouse completes the rectangle. Add other
  rectangles as needed. A single click creates a rectangle of default size
  centered on that click.
 
* **Add polygons**
  ![image: Add polygons](../../images/shape-add-polygons.png)

>   The `Add polygon` tool can be selected from the layer controls panel. Select

Reviewer: ```suggestion
  The `Add polygon` tool can be selected from the layer controls panel or by 
  using the default keybinding `p`. Select
```

Author: This section was also reformulated.

  the tool, click the location where you want the first vertex to be. Go to the
  second location and click there. Repeat until you have created all the
  vertices you want for this polygon. Hit the `Esc` key to stop creating the

>   polygon. This adds a final vertex at the current mouse position and completes

Reviewer: ```suggestion
  polygon or double-click, which will add a final vertex at the current mouse position and complete
```
```suggestion
  polygon. This adds a final vertex at the current mouse position and completes
```

Author: This section was reformulated.

![animation of napari viewer with a shapes layer bounded by vertices over an initial static image. The shapes are selected and the vertices moved by the mouse pointer, resulting in new shapes.](../../images/shape_vertex_editing.webm)

### Creating a new `shapes` layer

You can create a brand-new empty `shapes` layer by clicking the
`New shapes layer` button at the top of the `layer list` panel. The shape of the
new layer is defined by the shapes inside it, as new shapes are added the new
shape layer will adjust as needed. The dimension of the new shapes layer will
default to the largest dimension of any layer currently in the viewer, or to 2
if no other layers are present in the viewer.

### `shapes` layer buttons and controls

Here are the buttons and controls that are available in the GUI for the `shapes`
layer. This is the order they are presented in, not necessarily the order they
are used. i.e. You can't remove a vertex before you have created a shape.

* Buttons
    * Remove vertex
    * Insert vertex
    * Delete selected shape
    * Select vertices
    * Select shapes
    * Pan/zoom
    * Move to back
    * Move to front
    * Add ellipses
    * Add rectangles

>     * Add polygons
>     * Add lines

Author: ```suggestion
    * Add polygons
    * Add polygons lasso
    * Add lines
```

For more information about layers, refer to [Layers at a glance](../../guides/layers).

## When to use the `shapes` layer

The `shapes` layer allows you to display a list of an NxD arrays, where each
array corresponds to one shape, specified by N points in D coordinates. You can
adjust the position, size, face color, edge color, and opacity of all the shapes
independently, both programmatically and from the GUI.

## Controlling the `shapes` layer using the GUI

Before we start talking about all the controls on the GUI, here is an overview
of how to handle the `shapes` layer.

### Selecting, resizing, moving, editing, and deleting shapes

All shapes are edited in the same way:
1. Click the select shapes tool.
2. Draw a box around the shape you want to edit.
3. Adjust the size or contour of the shape using the square handles that appear
   on the bounding box of the shape. When resizing a layer, hold down the
   `shift` key to lock the aspect ratio of the shape. Then you can continue to
   resize the shape with a fixed aspect ratio. Note: you have to hold the
   `shift` key down and when finished release the mouse button first! Here is a
   shape being resized:

>    ![image: shape resizing](../../images/shape_resizing.webm)

Reviewer: This video has sound

Author: lol, sorry about that! 🤦🏻‍♀️

5. Change the edge width of a shape or the width of a line or path by clicking
   the circle next to `edge width:` and dragging it to a new width.
6. Move the shape by dragging it.
7. Rotate the shape by clicking and dragging on the rotation handle above the
   shape bounding box.
8. Select multiple shapes by continuing to `shift`+click additional shapes after
   the first, or drag a box around the shapes to select.
9. Select all the shapes in the current slice by clicking the `a` key if you are
   in select mode. Once selected you can delete the shapes by clicking the
   delete button in the layer controls panel or pressing the delete key on the
   keyboard.

### Copying and pasting shapes

Copy and paste any selected shapes using the `ctrl-c` and `ctrl-v` keybindings
respectively. If you have a multidimensional shapes layer you can copy shapes
from one slice to another by pasting them into the new slice. The coordinates of
the shapes in the visible dimensions will be in the same place on the new slice
as in the old slice, but the rest of the coordinates will be updated with the
new slice values.

### Adding (inserting), editing, and deleting (removing) individual vertices

> ![animation of napari viewer with a shapes layer bounded by vertices over an initial static image. The shapes are selected and the vertices moved by the mouse pointer, resulting in new shapes.](../../images/shape_vertex_editing.webm)

Reviewer: this video has sound

viewer.close()
```

### Arguments of `view_shapes` and `add_shapes`

{meth}`~napari.view_layers.view_shapes` and {meth}`~napari.Viewer.add_shapes`
accept the same layer-creation parameters.

```{code-cell} python
:tags: [hide-output]

help(napari.view_shapes)

> ```

Reviewer: Same comment as before regarding this cell with `help()`

`layer.current_edge_width` property that will determine the edge width of the
next shape to be added or any currently selected shapes.

### Layer ordering

You can get the ordering of all the shapes using the `layer.z_indices` property.
You can also set the initial ordering of shapes by passing a list to the
`layer.z_index` property.

### `shapes` layer opacity

The {ref}`opacity value <layer_opacity>` applies to all shapes. You can
initialize the shape opacities using the `opacity` keyword argument which
accepts a single opacity value that will be applied globally. You can then
access the opacity using the `layer.opacity` property. In order to adjust the
opacity of individual shapes you need to adjust the alpha value in the
`layer.edge_color` and `layer.face_color` properties.

### Putting it all together

Here you can see an example of adding, selecting, and editing shapes and
changing their properties:

> ![image: editing shapes](../../images/editing_shapes.webm)

Reviewer: This video is also visibly dated, because colors are set using a color picker now.

viewer.close()
```

### Non-editable mode

To disable editing of the shapes layer, set the `editable` property of the layer
to `False`.

As noted in [3D rendering](#3d-rendering), below, the shapes layer is not
editable when using 3D rendering.

### 3D rendering

All layers can be rendered in both 2D and 3D.

The number of dimensions sliders will be 2 or 3 less than the total number of
dimensions of the layer, allowing you to browse volumetric timeseries data and
other high dimensional data. See [](../../gallery/nD_shapes) to see shapes in
both 2D and 3D:

> ![image: nD shapes](../../images/nD_shapes.webm)

Reviewer: This is an older version of napari, but not a big deal in my opinion it's just showing rendering.

Author: Yeah, I thought this could be refreshed some other time but now that you mention it I'll update too as long as I'm here.

Also note that for a multidimensional shape to be displayed on a given view
slice, all of its non-displayed coordinates must match the coordinates of that
view slice, i.e. the shape must be entirely defined within that plane.

For paths that are defined by coordinates spanning more than two dimensions, it
is possible to visualize them as 3D cylinders, see [](../../gallery/3D_paths)
for examples.

> ![image: 3D paths](../../images/3D_paths.png)

Reviewer: This is an older version of napari, but not a big deal in my opinion it's just showing rendering.
This image could be generated programmatically using the gallery example snippet?

Author: I will check out the autogeneration later, for now I just updated the screenshot.

opacity of individual shapes you need to adjust the alpha value in the
`layer.edge_color` and `layer.face_color` properties.

### Putting it all together

Here you can see an example of adding, selecting, and editing shapes and
changing their properties:

> ![image: editing shapes](../../images/editing_shapes.webm)

Reviewer: This video has sound.
